src/main/scala/org/broadinstitute/dsde/firecloud/service/ExportEntitiesByTypeActor.scala

```diff
@@ -1,79 +1,272 @@
+    val tempMembershipFile: File = File.newTemporaryFile(prefix = "membership_")
+    val membershipSink: Sink[ByteString, Future[IOResult]] = FileIO.toPath(tempMembershipFile.path)
+
+    // Headers
+    val entityHeaders: IndexedSeq[String] = TSVFormatter.makeEntityHeaders(entityType, metadata.attributeNames, attributeNames)
+    val membershipHeaders: IndexedSeq[String] = TSVFormatter.makeMembershipHeaders(entityType)
+
+    // Run the Split Entity Flow that pipes entities through the two flows to the two file sinks
+    // Result of this will be a tuple of Future[IOResult] that represents the success or failure of
+    // streaming content to the file sinks.
+    val fileStreamIOResults: (Future[IOResult], Future[IOResult]) = {
+      RunnableGraph.fromGraph(GraphDSL.create(entitySink, membershipSink)((_, _)) { implicit builder =>
+        (eSink, mSink) =>
+          import GraphDSL.Implicits._
+
+          // Sources
+          val entitySource: Outlet[Seq[Entity]] = builder.add(Source.fromFuture(entityBatches)).out
+          val entityHeaderSource: Outlet[ByteString] = builder.add(Source.single(ByteString(entityHeaders.mkString("\t") + "\n"))).out
+          val membershipHeaderSource: Outlet[ByteString] = builder.add(Source.single(ByteString(membershipHeaders.mkString("\t") + "\n"))).out
+
+          // Flows
+          val splitter: UniformFanOutShape[Seq[Entity], Seq[Entity]] = builder.add(Broadcast[Seq[Entity]](2))
+          val entityFlow: FlowShape[Seq[Entity], ByteString] = builder.add(Flow[Seq[Entity]].map { entities =>
+            val rows = TSVFormatter.makeEntityRows(entityType, entities, entityHeaders)
+            ByteString(rows.map { _.mkString("\t")}.mkString("\n") + "\n")
+          })
+          val membershipFlow: FlowShape[Seq[Entity], ByteString] = builder.add(Flow[Seq[Entity]].map { entities =>
+            val rows = TSVFormatter.makeMembershipRows(entityType, entities)
+            ByteString(rows.map { _.mkString("\t")}.mkString("\n") + "\n")
+          })
+          val eConcat: UniformFanInShape[ByteString, ByteString] = builder.add(Concat[ByteString]())
+          val mConcat: UniformFanInShape[ByteString, ByteString] = builder.add(Concat[ByteString]())
+
+          // Graph
+          entityHeaderSource                         ~> eConcat
+          entitySource ~> splitter ~> entityFlow     ~> eConcat ~> eSink
+          membershipHeaderSource                     ~> mConcat
+                          splitter ~> membershipFlow ~> mConcat ~> mSink
+          ClosedShape
+      }).run()
+    }
+
+    // Check that each file is completed
+    val fileStreamResult = for {
+      eResult <- fileStreamIOResults._1
+      mResult <- fileStreamIOResults._2
+    } yield eResult.wasSuccessful && mResult.wasSuccessful
+
+    // And then map those files to a ZIP.
+    val zipResult = fileStreamResult flatMap { s =>
+      if (s) {
+        val zipFile: Future[File] = writeFilesToZip(tempEntityFile, tempMembershipFile)
```

Comment:
There are two cases - singular and set types. 
* For singular entity downloads, we do start streaming right away since we don't have to do any file operations. This case is where we can avoid timeout problems with large sets. 
* For set types, we can't stream directly since we need to write files first. Once those files are written, we can then send them. I'm pretty sure there's no way to both start writing the files *AND* stream them at the same time. This case does *not* allow us to avoid timeouts. However, timeouts are not really a problem in this case because set downloads are generally *much* smaller (orders of magnitude smaller) than singular entities.